virtual_rainforest/core/base_model.py

```diff
@@ -119,34 +177,96 @@ def cleanup(self) -> None:
 
     @classmethod
     @abstractmethod
-    def from_config(cls, config: dict[str, Any]) -> Any:
+    def from_config(cls, data: Data, config: dict[str, Any]) -> BaseModel:
         """Factory function to unpack config and initialise a model instance."""
 
     @classmethod
     def __init_subclass__(cls) -> None:
-        """Method that adds new model classes to the model registry.
+        """Initialise subclasses deriving from BaseModel.
```

Comment:
The functionality of this method is totally ok, but I would split it into smaller functions for checking the individual attributes, leaving this as a top level method that just call those internally. That will make understanding what's going here on and testing easier.

Reply:
This was _much_ more irritating to do than anticipated. `mypy` basically complained about _everything_ once the components of the type checking (which is essentially the whole of the function) were isolated in separate functions. And at the end, we now have `# type: ignore`s because `mypy` in `__init_subclass` is no longer aware that the properties have been checked and have the correct structure.

Sometimes I _hate_ `mypy`.

Reply:
I mean, I agree that the end point is cleaner, but.... gaaaaaah!